src/Microsoft.AspNetCore.Mvc.Core/Routing/MvcEndpointFinder.cs

```diff
@@ -0,0 +1,254 @@
+            if (address.TargetActionDescriptor != null)
+            {
+                return FindEndpointsByActionDescriptor(address.TargetActionDescriptor);
+            }
+
+            var linkTarget = GetLinkTarget(address);
+
+            ActionDescriptor targetActionDescriptor = null;
+            if (linkTarget == LinkTarget.Mvc)
+            {
+                targetActionDescriptor = GetTargetControllerActionDescriptor(address);
+            }
+            else if (linkTarget == LinkTarget.RazorPages)
+            {
+                targetActionDescriptor = GetTargetPageActionDescriptor(address);
+            }
+
+            if (targetActionDescriptor == null)
+            {
+                return Empty;
+            }
+
+            return FindEndpointsByActionDescriptor(targetActionDescriptor);
+        }
+
+        private ActionDescriptor GetTargetControllerActionDescriptor(MvcAddress address)
+        {
+            ActionDescriptor targetActionDescriptor = null;
+            // Same controller, different action
+            if (!string.IsNullOrEmpty(address.TargetActionName) &&
+                string.IsNullOrEmpty(address.TargetControllerName))
+            {
+                var currentActionDescriptor = (ControllerActionDescriptor)address.CurrentActionContext.ActionDescriptor;
+                var currentControllerType = currentActionDescriptor.ControllerTypeInfo;
+
+                targetActionDescriptor = _actionDescriptorCollectionProvider.ActionDescriptors.Items
+                    .OfType<ControllerActionDescriptor>()
+                    .FirstOrDefault(ad => ad.ControllerTypeInfo.Equals(currentControllerType) &&
+                    string.Equals(ad.ActionName, address.TargetActionName, StringComparison.OrdinalIgnoreCase));
+            }
+            // Different controller, different action
+            else if (!string.IsNullOrEmpty(address.TargetActionName) &&
+                !string.IsNullOrEmpty(address.TargetControllerName))
+            {
+                targetActionDescriptor = _actionDescriptorCollectionProvider.ActionDescriptors.Items
+                    .OfType<ControllerActionDescriptor>()
+                    .FirstOrDefault(
+                    ad => string.Equals(ad.ControllerName, address.TargetControllerName, StringComparison.OrdinalIgnoreCase) &&
+                    string.Equals(ad.ActionName, address.TargetActionName, StringComparison.OrdinalIgnoreCase));
+            }
+            return targetActionDescriptor;
+        }
+
+        private ActionDescriptor GetTargetPageActionDescriptor(MvcAddress address)
+        {
+            if (!string.IsNullOrEmpty(address.TargetPageName))
+            {
+                var targetActionDescriptor = _actionDescriptorCollectionProvider.ActionDescriptors.Items
+                    .FirstOrDefault(ad =>
+                    {
+                        if (ad.RouteValues.TryGetValue("page", out var pageName) &&
```

Comment:
I do not like this but since this url helper sits in Mvc.Core, I cannot access PageActionDesriptor from here